File: ios/PreferenceModule/Shared/Plugin/PreferencePlugin.m
//
//  PreferencePlugin.m
//  PreferenceModule
//
//  Created by JoAmS on 2017. 6. 14..
//  Copyright © 2017년 t5online. All rights reserved.
//

#import "PreferencePlugin.h"
#import <NebulaCore/Nebula.h>
#import "PreferenceService.h"

#define ERROR_CODE_INVALID_KEY      @"E10001"
#define ERROR_CODE_INVALID_VALUE    @"E10002"
#define ERROR_CODE_INVALID_DATA     @"E10003"

@implementation PreferencePlugin

- (void)get:(NSString*)key defaultValue:(NSString*)defaultValue
{
    if (!key) {
        [self reject:ERROR_CODE_INVALID_KEY message:@"Invalid Key" data:nil];
        return;
    }
    
    NSString* ret = [[Nebula serviceWithKey:SERVICE_KEY_PREFERENCE] getPreference:key defaultValue:defaultValue];
    
    NSMutableDictionary* retData = [NSMutableDictionary dictionary];
    [retData setObject:ret forKey:@"value"];
    [self resolve:retData];
    
}

- (void)set:(NSString*)key value:(NSString*)value
{
    if (!key) {
        [self reject:ERROR_CODE_INVALID_KEY message:@"Invalid Key" data:nil];
        return;
    }
    
    if (!value) {
        [self reject:ERROR_CODE_INVALID_VALUE message:@"Invalid Value" data:nil];
        return;
    }
    
    [[Nebula serviceWithKey:SERVICE_KEY_PREFERENCE] setPreference:value forKey:key];
    [self resolve];
}

- (void)remove:(NSString *)key
{
    if (!key) {
        [self reject:ERROR_CODE_INVALID_KEY message:@"Invalid Key" data:nil];
        return;
    }
    
    [[Nebula serviceWithKey:SERVICE_KEY_PREFERENCE] removePreference:key];
    [self resolve];
}

- (void)removeAll
{
    [[Nebula serviceWithKey:SERVICE_KEY_PREFERENCE] removeAllPreference];
    [self resolve];
}

@end

File: android/PreferenceModule/app/src/main/java/shared/plugin/PreferencePlugin.java
package shared.plugin;

import com.t5online.nebulacore.Nebula;
import com.t5online.nebulacore.plugin.Plugin;

import org.json.JSONException;
import org.json.JSONObject;

import shared.service.PreferenceService;

/**
 * Created by JoAmS on 2017. 6. 15..
 */

public class PreferencePlugin extends Plugin {

    public static final String PLUGIN_GROUP_PREFERENCE = "preference";

    public void get(String key, String defaultValue) {
        String value = "";
        try {
            PreferenceService preferenceService = (PreferenceService) Nebula.getService(PreferenceService.SERVICE_KEY_PREFERENCE);
            value = preferenceService.getPreference(key, defaultValue);
            JSONObject ret = new JSONObject();
            ret.put("code", STATUS_CODE_SUCCESS);
            ret.put("message", value);
            resolve(ret);
        } catch (JSONException e) {
            e.printStackTrace();
            reject();
        }
    }

    public void set(String key, String value) {
        PreferenceService preferenceService = (PreferenceService) Nebula.getService(PreferenceService.SERVICE_KEY_PREFERENCE);
        preferenceService.setPreference(key, value);
        try {
            JSONObject ret = new JSONObject();
            ret.put("code", STATUS_CODE_SUCCESS);
            ret.put("message", "");
            resolve(ret);
        }catch (Exception e){
            e.printStackTrace();
            reject();
        }
    }

    public void remove(String key) {
        PreferenceService preferenceService = (PreferenceService) Nebula.getService(PreferenceService.SERVICE_KEY_PREFERENCE);
        preferenceService.removePreference(key);
        try {
            JSONObject ret = new JSONObject();
            ret.put("code", STATUS_CODE_SUCCESS);
            ret.put("message", "");
            resolve(ret);
        }catch (Exception e){
            e.printStackTrace();
            reject();
        }
    }

    public void removeAll() {
        PreferenceService preferenceService = (PreferenceService) Nebula.getService(PreferenceService.SERVICE_KEY_PREFERENCE);
        preferenceService.removeAllPreference();
        try {
            JSONObject ret = new JSONObject();
            ret.put("code", STATUS_CODE_SUCCESS);
            ret.put("message", "");
            resolve(ret);
        }catch (Exception e){
            e.printStackTrace();
            reject();
        }
    }
}
